angr/state_plugins/symbolizer.py

```diff
@@ -0,0 +1,252 @@
+            s = self._preconstrain(le).reversed
+            l.debug("Replacing %#x (at %#x, endness LE) with %s!", le, base+offset, s)
+            return s
+        else:
+            return None
+
+    def _resymbolize_data(self, data, prefix=b"", base=0, skip=()):
+        ws = self.state.arch.bytes
+        suffix = data[len(data)-(len(data)%ws):]
+        data = data[:len(data)-(len(data)%ws)]
+
+        num_words = len(data) // ws
+        unpacked_le = struct.unpack(self._LE_FMT[0] + str(num_words) + self._LE_FMT[1], data)
+        unpacked_be = struct.unpack(self._BE_FMT[0] + str(num_words) + self._BE_FMT[1], data)
+
+        values_squashed = [ prefix ]
+        last_idx = 0
+        for i,(be,le) in enumerate(zip(unpacked_be, unpacked_le)):
+            #assert len(claripy.Concat(*values_squashed)) == i*8
+
+            s = self._resymbolize_int(be, le, base, i*ws, skip)
+            if s is None:
+                return None
+
+            if last_idx != i:
+                values_squashed.append(data[last_idx*ws:i*ws])
+            last_idx = i + 1
+            values_squashed.append(s)
+
+        if len(values_squashed) == 1:
+            return None
+
+        if last_idx != num_words:
+            values_squashed.append(data[last_idx*ws:])
+        values_squashed.append(suffix)
+
+        new_data = claripy.Concat(*values_squashed)
+        #assert len(new_data)/8 == len(data) + len(prefix)
+        #assert self.state.solver.eval_one(new_data) == self.state.solver.eval_one(claripy.BVV(data))
+        return new_data
+
+    def _resymbolize_region(self, storage, addr, length):
+        assert type(addr) is int
+        assert type(length) is int
+
+        self.state.scratch.push_priv(True)
+        memory_objects = storage.mem.load_objects(addr, length)
+        self.state.scratch.pop_priv()
+
+        for _,mo in memory_objects:
+            if not mo.is_bytes and mo.object.symbolic:
+                l.debug("Skipping symbolic memory object %s.", mo)
+                continue
+
+            aligned_base = mo.base - mo.base % (-self.state.arch.bytes)
+            remaining_len = mo.last_addr + 1 - aligned_base
+            if remaining_len < self.state.arch.bytes:
+                continue
+
+            data = mo.bytes_at(aligned_base, remaining_len, allow_concrete=True)
+            if not mo.is_bytes:
+                data = self.state.solver.eval_one(data, cast_to=bytes)
+            #assert self.state.solver.eval_one(storage.load(aligned_base, remaining_len, endness='Iend_BE'), cast_to=bytes).rjust(self.state.arch.bytes) == data
+
+            replacement_content = self._resymbolize_data(
+                data,
+                base=aligned_base,
+                prefix=mo.bytes_at(mo.base, mo.length - remaining_len) if aligned_base != mo.base else b"",
+                skip=() if storage is self.state.memory else (self.state.arch.ip_offset,)
```

Comment:
This is the only use of skip and it's for a single offset only, there should be a better way of handling this I think.